rfcs/2021-11-12-9811-vrl-vm.md

```diff
@@ -0,0 +1,347 @@
+result on the stack. Every operation pulls the values it uses from the stack.
+
+- parameter_stack
+
+For function calls, the parameters need to be evaluated and passed to the
+function. The stack is a `Vec` of parameter name and value. We need an OpCode
+that will copy a value from the main stack to the parameter stack tagged with
+the parameter name. The `FunctionArgs` compiler will dump this OpCode after the
+code to evaluate the given arg is dumped.
+
+- error
+
+If an operation errors, it needs to populate this field. Operations such as
+the error assignment (`result, error = thing()`) will need to check if this
+field has been populated and action accordingly.
+
+- instruction_pointer
+
+The instruction pointer points to the next instruction to evaluate.
+
+#### Calling functions
+
+Calling functions in the stdlib will be a case of evaluating each parameter
+with the results pushed onto the parameter stack.
+
+Since VRL allows for named parameters, parameters need to be pushed on to the
+parameter stack _in the order that they are specified in the `Function`
+implementation_.
+
+Parameters are optional and may not be specified. However, to avoid passing
+parameters to the wrong function, an OpCode still needs to be emitted to move
+a placeholder value - `None` - to the parameter stack.
+
+Take the following hypothetical situation:
+
+```coffeescript
+thing(surname: "nong", name: thang(name: "nork"))
+```
+
+Supposing function `thang` also took an optional parameter `surname`, when
+`thang` is called, we will have a `surname` parameter on the parameter stack.
+Without the placeholder, `thang` would thing that `surname` was being passed to
+it. If the placeholder is pushed, it would consume this instead.
+
+We do not want `stdlib` functions to have access to the VM since that risks a
+rogue function destroying the VM's state.
+
+Current `stdlib` functions are composed of a combination of `compile` and
+`resolve` functions. These function will need to be combined into a single
+function `call`.
+
+The `ArgumentList` parameter that is passed into `call` will have access to the
+parameter stack and the parameter list exposed by the `Function`. This can use
+these to return the appropriate values for the `required`, `optional` etc..
+functions.
+
+Since the `compile` function was also used by some `stdlib` functions to
+validate the parameters we may alse need a function `validate_parameters` that
+these functions can use to perform extra validation on the parameters.
+
+### Debugging
+
+We need debugging tools to help us introspect and debug the compiled bytecode.
+
+#### Dissassembler
+
+Output the compiled code in readable format to the console.
+
+#### Step debugger
+
+Output each step of the VM to the console, print the Stack and other state
+variables in the VM. Allow the user to press a key to move to the next step.
+
+#### Linter
+
+Check each instruction to ensure:
+
+- each Jump Opcode jumps to a valid location in the bytecode.
```

Comment:
Currently, given that there are no loops, also check that each jump is a forward jump. I understand that may be short lived, though.

Reply:
Theoretically we could create different OpCodes for forward and backward jumps which would still allow for this check.

Reply:
I would even go so far as to say VRL design should be kept so that we can maintain this check.